src/ehttpc.erl

```diff
@@ -633,20 +673,98 @@ enqueue_req(ReplyTo, Req, #state{requests = Requests0} = State) ->
     State#state{requests = drop_expired(Requests)}.
 
 %% call gun to shoot the request out
-maybe_shoot(#state{enable_pipelining = EP, requests = Requests0, client = Client} = State0) ->
-    #{sent := Sent} = Requests0,
+maybe_shoot(
+    #state{
+        requests =
+            #{
+                sent := Sent,
+                max_sent_expire := MaxExpire
+            } = Requests0,
+        client = Client,
+        max_inactive = MaxInactive,
+        enable_pipelining = PipelineLimit
+    } = State0
+) ->
     State = State0#state{requests = drop_expired(Requests0)},
-    %% If the gun http client is down
-    ClientDown = is_pid(Client) andalso (not is_process_alive(Client)),
-    %% Or when too many has been sent already
-    case ClientDown orelse should_cool_down(EP, maps:size(Sent)) of
-        true ->
+    SentCount = map_size(Sent),
+    case check_gun(Client, PipelineLimit, SentCount, MaxExpire, MaxInactive) of
+        continue ->
+            do_shoot(State);
+        pause ->
             %% Then we should cool down, and let the gun responses
             %% or 'EXIT' message to trigger the flow again
-            ?tp(cool_down, #{enable_pipelining => EP}),
+            ?tp(cool_down, #{enable_pipelining => State#state.enable_pipelining}),
             State;
+        reconnect ->
+            %% the connection has been inactive for too long
+            log(
+                error,
+                #{
+                    msg => "force_reconnecting_zombie_http_connection",
+                    last_request_expire => calendar:system_time_to_rfc3339(MaxExpire, [
```

Comment:
Nit: add only if `MaxExpire > 0`, or set it to `undefined` otherwise?

Reply:
it's not possible to end up in `reconnect` if `MaxExpire` is 0.
will add an assertion here.